cmd2/cmd2.py

```diff
@@ -381,6 +399,229 @@ def __init__(self, completekey: str = 'tab', stdin=None, stdout=None, *,
+        except Exception:
+            for attrib in installed_attributes:
+                delattr(self, attrib)
+            if cmdset in self._installed_command_sets:
+                self._installed_command_sets.remove(cmdset)
+            raise
+
+    def _install_command_function(self, command: str, command_wrapper: Callable, context=''):
+        cmd_func_name = COMMAND_FUNC_PREFIX + command
+
+        # Make sure command function doesn't share naem with existing attribute
+        if hasattr(self, cmd_func_name):
+            raise ValueError('Attribute already exists: {} ({})'.format(cmd_func_name, context))
+
+        # Check if command has an invalid name
+        valid, errmsg = self.statement_parser.is_valid_command(command)
+        if not valid:
+            raise ValueError("Invalid command name {!r}: {}".format(command, errmsg))
+
+        # Check if command shares a name with an alias
+        if command in self.aliases:
+            self.pwarning("Deleting alias '{}' because it shares its name with a new command".format(command))
+            del self.aliases[command]
+
+        # Check if command shares a name with a macro
+        if command in self.macros:
+            self.pwarning("Deleting macro '{}' because it shares its name with a new command".format(command))
+            del self.macros[command]
+
+        setattr(self, cmd_func_name, command_wrapper)
+
+    def _install_completer_function(self, cmd_name: str, cmd_completer: Callable):
+        completer_func_name = COMPLETER_FUNC_PREFIX + cmd_name
+
+        if hasattr(self, completer_func_name):
+            raise ValueError('Attribute already exists: {}'.format(completer_func_name))
+        setattr(self, completer_func_name, cmd_completer)
+
+    def _install_help_function(self, cmd_name: str, cmd_help: Callable):
+        help_func_name = HELP_FUNC_PREFIX + cmd_name
+
+        if hasattr(self, help_func_name):
+            raise ValueError('Attribute already exists: {}'.format(help_func_name))
+        setattr(self, help_func_name, cmd_help)
+
+    def uninstall_command_set(self, cmdset: CommandSet):
+        """
+        Uninstalls a CommandSet and unloads all associated commands
+        :param cmdset: CommandSet to uninstall
+        """
+        if cmdset in self._installed_command_sets:
+
+            self._unregister_subcommands(cmdset)
+
+            methods = inspect.getmembers(
+                cmdset,
+                predicate=lambda meth: isinstance(meth, Callable)
+                and hasattr(meth, '__name__') and meth.__name__.startswith(COMMAND_FUNC_PREFIX))
+
+            for method in methods:
+                cmd_name = method[0][len(COMMAND_FUNC_PREFIX):]
+
+                # Enable the command before uninstalling it to make sure we remove both
+                # the real functions and the ones used by the DisabledCommand object.
+                if cmd_name in self.disabled_commands:
+                    self.enable_command(cmd_name)
```

Comment:
Recommend adding a unit test for the case where the code within this `if` is it because the command is in `disabled_commands`